src/CodeGen_RISCV.cpp

```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+    if (effective_vscale != 0) {
+        for (const RISCVIntrinsic &intrin : intrinsic_defs) {
+            if (intrin.ret_type.type_pattern == FixedOrScalableVector::WildcardWidths) {
+                for (int scale = 0; scale < 4; scale++) {
+                    int bit_width_scale = 1 << scale;
+
+                    Type ret_type = concretize_fixed_or_scalable(intrin.ret_type, bit_width_scale,
+                                                                 target.vector_bits);
+
+                    std::vector<Type> arg_types;
+                    if ((intrin.ret_type.relative_scale * bit_width_scale * intrin.ret_type.type.bits()) > 64) {
+                        break;
+                    }
+
+                    // Add an initial argument to handle tail propagation.
+                    arg_types.push_back(ret_type);
+
+                    for (const auto &arg_type : intrin.arg_types) {
+                        if (arg_type.type_pattern == FixedOrScalableVector::Undefined) {
+                            break;
+                        }
+                        if ((arg_type.relative_scale * bit_width_scale * arg_type.type.bits()) > 64) {
+                            break;
+                        }
+                        arg_types.push_back(concretize_fixed_or_scalable(arg_type, bit_width_scale,
+                                                                         target.vector_bits));
+                    }
+                    llvm::Function *intrin_impl = define_riscv_intrinsic_wrapper(intrin, bit_width_scale);
+                    declare_intrin_overload(intrin.name, ret_type, intrin_impl, arg_types);
+                }
+            } else {
+                llvm::Function *intrin_impl = define_riscv_intrinsic_wrapper(intrin, 1);
+                Type ret_type = concretize_fixed_or_scalable(intrin.ret_type, 1,
+                                                             target.vector_bits);
+                std::vector<Type> arg_types;
+
+                // Add an initial argument to handle tail propagation.
+                arg_types.push_back(ret_type);
+
+                for (const auto &arg_type : intrin.arg_types) {
+                    if (arg_type.type_pattern == FixedOrScalableVector::Undefined) {
+                        break;
+                    }
+                    arg_types.push_back(concretize_fixed_or_scalable(arg_type, 1, target.vector_bits));
+                }
+                declare_intrin_overload(intrin.name, ret_type, intrin_impl, arg_types);
+            }
+        }
+    }
+}
+
+llvm::Function *CodeGen_RISCV::define_riscv_intrinsic_wrapper(const RISCVIntrinsic &intrin,
+                                                              int bit_width_scale) {
+    int effective_vscale = target_vscale();
+
+    llvm::Type *xlen_type = target.bits == 32 ? i32_t : i64_t;
+
+    // Produce intrinsic name and
+    std::vector<llvm::Type *> llvm_arg_types;
+    std::string mangled_name = "llvm.riscv.";
```

Comment:
Because that will not compile without casting one of the ```const char *``` values to ```std::string``` and if I have to mention the type again anyway, this strikes me as a bit clearer.

Reply:
Gotcha